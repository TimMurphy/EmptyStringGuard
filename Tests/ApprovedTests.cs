#if(DEBUG)
using Microsoft.Build.Utilities;
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using ApprovalTests;
using ApprovalTests.Reporters;
using NUnit.Framework;

[Ignore("ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40) returns null on my machine")]
[TestFixture]
[UseReporter(typeof(DiffReporter), typeof(ClipboardReporter))]
public class ApprovedTests
{
    [Test]
    public void ClassWithBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithBadAttributes"));
    }

    [Test]
    public void ClassWithPrivateMethod()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "ClassWithPrivateMethod"));
    }

    [Test]
    public void InterfaceBadAttributes()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "InterfaceBadAttributes"));
    }

    [Test]
    public void SimpleClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SimpleClass"));
    }

    [Test]
    public void SkipIXamlMetadataProvider()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "XamlMetadataProvider"));
    }

    [Test]
    public void SpecialClass()
    {
        Approvals.Verify(Decompile(AssemblyWeaver.AfterAssemblyPath, "SpecialClass"));
    }

    private static string Decompile(string assemblyPath, string identifier = "")
    {
        var exePath = GetPathToILDasm();

        if (!string.IsNullOrEmpty(identifier))
            identifier = "/item:" + identifier;

        using (var process = Process.Start(new ProcessStartInfo(exePath, $"\"{assemblyPath}\" /nobar /text {identifier}")
        {
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        }))
        {

            var projectFolder = Path.GetFullPath(Path.GetDirectoryName(TestContext.CurrentContext.TestDirectory) + @"\..\..").Replace(@"\", @"\\");
            projectFolder = $@"{projectFolder[0]}{projectFolder.Substring(1)}\\";

            process.WaitForExit(10000);

            var decompile = process.StandardOutput.ReadToEnd()
                .Replace(projectFolder, "");

            return decompile.Substring(decompile.IndexOf("// warning :"));
        }
    }

    private static string GetPathToILDasm()
    {
        var path = Path.Combine(ToolLocationHelper.GetPathToDotNetFrameworkSdk(TargetDotNetFrameworkVersion.Version40), @"bin\NETFX 4.0 Tools\ildasm.exe");
        if (!File.Exists(path))
            path = path.Replace("v7.0", "v8.0");
        if (!File.Exists(path))
            Assert.Ignore("ILDasm could not be found");
        return path;
    }
}

#endif