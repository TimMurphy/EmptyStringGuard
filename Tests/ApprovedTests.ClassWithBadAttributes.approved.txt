// warning : THIS IS A PARTIAL DISASSEMBLY, NOT SUITABLE FOR RE-ASSEMBLING



.class public abstract auto ansi beforefieldinit ClassWithBadAttributes
       extends [mscorlib]System.Object
{
  .method public hidebysig newslot abstract virtual 
          instance void  MethodWithEmptyStringCheckOnParam(string arg) cil managed
  {
  } // end of method ClassWithBadAttributes::MethodWithEmptyStringCheckOnParam

  .method family hidebysig specialname rtspecialname 
          instance void  .ctor() cil managed
  {
    // Code size       7 (0x7)
    .maxstack  8
    IL_0000:  ldarg.0
    IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
    IL_0006:  ret
  } // end of method ClassWithBadAttributes::.ctor

} // end of class ClassWithBadAttributes

// *********** DISASSEMBLY COMPLETE ***********************
